core/classes/Misc/MentionsParser.php

```diff
@@ -11,49 +11,78 @@
+            $content,
+            $notificationRecipients,
+            $author_id,
+            null,
+            false,
+            $url,
+        );
+
+        $notification->send();
+
+        return self::replaceWithBbcode($content, $receipients);
+    }
+
+    /**
+     * Get users from the database based on the provided nicknames. Filters out users have blocked the author.
+     */
+    private static function getRecipients(string $content, int $author_id): array
+    {
+        preg_match_all(self::USER_MENTIONS_REGEX, $content, $matches);
+        $nicknames = $matches[1];
+
+        return DB::getInstance()->query(
+            'SELECT u.id, u.nickname FROM nl2_users u WHERE u.nickname IN (' . implode(',', array_map(static fn ($_) => '?', $nicknames)) . ') AND NOT EXISTS (SELECT 1 FROM nl2_blocked_users bu WHERE bu.user_id = u.id AND bu.user_blocked_id = ?)',
+            [
+                ...$nicknames,
+                $author_id,
+            ]
+        )->results();
+    }
+
     /**
-     * Parse the given HTML to include @username tags.
+     * Replace @username tags with [user]<id>[/user] bbcode.
      *
-     * @param int     $author_id   User ID of post creator.
-     * @param string  $value       Post content.
-     * @param ?string $link        Link back to post.
-     * @param ?array  $alert_short Short alert info, leave null to not alert user.
-     * @param ?array  $alert_full  Full alert info, leave null to not alert user.
+     * @param string $content     Post/custom page content.
+     * @param array  $receipients Array of user objects (with nickname and ID fields).
      *
      * @return string Parsed post content.
      */
-    public static function parse(int $author_id, string $value, ?string $link = null, ?array $alert_short = null, ?array $alert_full = null): string
+    private static function replaceWithBbcode(string $content, array $receipients): string
     {
-        if (preg_match_all('/(?<!\/)@([A-Za-z0-9\-_!.]+)/', $value, $matches)) {
-            $matches = $matches[1];
-
-            foreach ($matches as $possible_username) {
-                $user = null;
-
-                while (($possible_username != '') && !$user) {
-                    $user = new User($possible_username, 'nickname');
-
-                    if ($user->exists()) {
-                        $value = preg_replace('/(?<!\/)' . preg_quote("@$possible_username", '/') . '/', '[user]' . $user->data()->id . '[/user]', $value);
-
-                        // Check if user is blocked by OP
-                        if ($value && $link && ($alert_full && $alert_short) && ($user->data()->id != $author_id) && !$user->isBlocked($user->data()->id, $author_id)) {
-                            Alert::create($user->data()->id, 'tag', $alert_short, $alert_full, $link);
-                            break;
-                        }
-                    }
-
-                    // chop last word off of it
```

Comment:
Why is that not needed anymore?

Reply:
I couldn't figure out why it was needed initially. Do you know what it does?